### PR TITLE
MINOR: Move KTable source topic for changelog to optimization framework

### PR DESCRIPTION
**Pull request by bbejeck**
Since we've added Kafka Streams optimizations in `2.1` we need to move the optimization for source `KTable` nodes (use source topic as changelog) to the optimization framework.

Updated streams tests.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by bbejeck** [comment]
ping @guozhangwang @mjsax @vvcephei @ableegoldman for reviews.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableSourceNode.java [hunk 1]
```diff
@@ -104,7 +110,11 @@ public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
             final KTableSource<K, V> ktableSource = (KTableSource<K, V>) processorParameters.processorSupplier();
             if (ktableSource.queryableName() != null) {
                 topologyBuilder.addStateStore(storeBuilder, nodeName());
-                topologyBuilder.markSourceStoreAndTopic(storeBuilder, topicName);
+
+                if (shouldReuseSourceTopicForChangelog) {
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableSourceNode.java:114]: This is the optimization and we only apply it for materialized `KTable` objects.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableSourceNode.java [hunk 2]
```diff
@@ -103,8 +109,13 @@ public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
...
             final KTableSource<K, V> ktableSource = (KTableSource<K, V>) processorParameters.processorSupplier();
             if (ktableSource.queryableName() != null) {
-                topologyBuilder.addStateStore(storeBuilder, nodeName());
-                topologyBuilder.markSourceStoreAndTopic(storeBuilder, topicName);
+                if (shouldReuseSourceTopicForChangelog) {
+                    storeBuilder.withLoggingDisabled();
+                    topologyBuilder.addStateStore(storeBuilder, nodeName());
+                    topologyBuilder.connectSourceStoreAndTopic(storeBuilder.name(), topicName);
+                } else {
+                    topologyBuilder.addStateStore(storeBuilder, nodeName());
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableSourceNode.java]: nit: we can move this line up outside the if condition and remove else?

**Review by guozhangwang — APPROVED** [review]
One nit, otherwise LGTM.

Feel free to merge afterwards.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableSourceNode.java (hunk 2 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/TableSourceNode.java]: Ack

**Comment by bbejeck** [comment]
Addressed comments and rebased with trunk, waiting for Jenkins to complete for merging

**Comment by bbejeck** [comment]
Merged #6500 into trunk.

**Pull request merged**